File: clients/android/app/src/main/java/id/my/anciety/codever/client/MatrixNativeProjection.kt
package id.my.anciety.codever.client

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put

/** Rebuilds the Web bridge's compatibility snapshot from Matrix thread events. */
class MatrixNativeProjection {
    private data class NativeRevision(
        val revision: Long,
        val epoch: String,
        val generation: Long,
    )

    private val sessions = linkedMapOf<String, JsonObject>()
    private val pendingDeltas = linkedMapOf<String, MutableList<JsonObject>>()
    private val deletedSessions = mutableMapOf<String, NativeRevision>()
    private var checkpoint: JsonObject? = null
    private var latestRevision: NativeRevision? = null

    /**
     * Restores the projection baseline cached by the Web bridge. The direct
     * Matrix receiver resumes from a persisted /sync cursor, so after a
     * process restart it may see an incremental root/lifecycle event before
     * another checkpoint. Seeding the projector keeps those live events
     * applicable without replaying old room history.
     */
    @Synchronized
    fun restore(snapshot: JsonObject): JsonObject {
        require(snapshot.number("version") == 1L)
        require(snapshot.text("kind") == "gateway_state")
        val revision = nativeRevision(snapshot)
        val stateVersion = snapshot.number("state_version") ?: error("Gateway state version is missing.")
        val activeDeviceCount = snapshot.number("active_device_count")
            ?: error("Gateway active device count is missing.")
        val workspace = snapshot["workspace"] as? JsonObject
            ?: error("Gateway workspace is missing.")
        val capabilities = snapshot["capabilities"] as? JsonObject
            ?: error("Gateway capabilities are missing.")
        val restoredSessions = snapshot["sessions"] as? JsonArray
            ?: error("Gateway sessions are missing.")

        sessions.clear()
        pendingDeltas.clear()
        deletedSessions.clear()
        restoredSessions.forEach { entry ->
            val session = entry as? JsonObject ?: error("Gateway session is invalid.")
            val sessionId = session.text("id") ?: error("Gateway session id is missing.")
            require(session.number("updated_at") != null)
            require(session.text("project_id") != null)
            require(session.text("project_name") != null)
            require(session.text("cwd") != null)
            require(session.text("provider") != null)
            sessions[sessionId] = session
        }
        checkpoint = buildJsonObject {
            put("version", 2)
            put("kind", "gateway_checkpoint")
            put("revision", revision.revision)
            put("revision_epoch", revision.epoch)
            put("revision_epoch_generation", revision.generation)
            put("state_version", stateVersion)
            put("active_device_count", activeDeviceCount)
            put("workspace", buildJsonObject {
                put("project", buildJsonObject {
                    put("id", workspace.text("project_id") ?: error("Workspace project id is missing."))
                    put("name", workspace.text("project_name") ?: error("Workspace project name is missing."))
                    put("cwd", workspace.text("cwd") ?: error("Workspace cwd is missing."))
                })
                put("provider", workspace.text("provider") ?: error("Workspace provider is missing."))
                workspace.text("model")?.let { put("model", it) }
                workspace.text("reasoning_effort")?.let { put("reasoning_effort", it) }
                put("permission_mode", workspace.text("permission_mode") ?: "default")
            })
            put("capabilities", capabilities)
        }
        latestRevision = revision
        return snapshot() ?: error("Gateway state projection could not be restored.")
    }

    @Synchronized
    fun apply(extension: JsonObject): JsonObject? {
        val kind = extension.text("kind")
        if (kind !in NATIVE_KINDS) return snapshot()
        observeRevision(extension)
        if (kind != "gateway_checkpoint" && isOlderThanCheckpoint(extension)) return snapshot()
        when (kind) {
            "gateway_checkpoint" -> {
                require(extension.number("version") == 2L)
                val current = checkpoint?.number("state_version") ?: -1
                if ((extension.number("state_version") ?: -1) > current) {
                    checkpoint = extension
                    applyCheckpointSessions(extension)
                }
            }
            "gateway_revision" -> Unit
            "session_root" -> applyRoot(extension)
            "session_update" -> applyUpdate(extension, queueIfMissing = true)
            "session_lifecycle" -> applyLifecycle(extension, queueIfMissing = true)
        }
        return snapshot()
    }

    @Synchronized
    fun applyStatus(extension: JsonObject): JsonObject? {
        val sessionId = extension.text("session_id") ?: return snapshot()
        val current = sessions[sessionId] ?: return snapshot()
        val state = extension.text("state").takeIf { it in setOf("running", "stopping", "failed") }
            ?: "idle"
        sessions[sessionId] = JsonObject(current + buildMap {
            put("status", JsonPrimitive(state))
            extension.text("activity_phase")?.let { put("activity_phase", JsonPrimitive(it)) }
        })
        return snapshot()
    }

    private fun applyCheckpointSessions(value: JsonObject) {
        val inventory = value["sessions"] as? JsonArray ?: return
        val checkpointRevision = nativeRevision(value)
        deletedSessions.entries.removeAll { (_, revision) ->
            compareRevisions(revision, checkpointRevision) < 0
        }
        pendingDeltas.entries.removeAll { (_, deltas) ->
            deltas.removeAll { compareRevisions(nativeRevision(it), checkpointRevision) < 0 }
            deltas.isEmpty()
        }
        sessions.clear()
        inventory.forEach { entry ->
            val session = entry as? JsonObject ?: return@forEach
            val sessionId = session.text("session_id") ?: return@forEach
            if (sessionId in deletedSessions) return@forEach
            val project = session["project"] as? JsonObject ?: return@forEach
            val updatedAt = session.number("updated_at") ?: return@forEach
            val projectId = project.text("id") ?: return@forEach
            val projectName = project.text("name") ?: return@forEach
            val cwd = project.text("cwd") ?: return@forEach
            val provider = session.text("provider") ?: return@forEach
            sessions[sessionId] = buildJsonObject {
                put("id", sessionId)
                put("title", session.text("title") ?: "Session")
                put("updated_at", updatedAt)
                put("status", session.text("status") ?: "idle")
                if (session["archived"]?.jsonPrimitive?.contentOrNull == "true") {
                    put("archived", true)
                }
                session.text("activity_phase")?.let { put("activity_phase", it) }
                put("project_id", projectId)
                put("project_name", projectName)
                put("cwd", cwd)
                put("provider", provider)
                session.text("model")?.let { put("model", it) }
                session.text("reasoning_effort")?.let { put("reasoning_effort", it) }
                put("extensions", session["extensions"] as? JsonArray ?: JsonArray(emptyList()))
            }
            drainPendingDeltas(sessionId)
        }
    }

    private fun applyRoot(value: JsonObject) {
        require(value.number("version") == 2L)
        val sessionId = value.text("session_id") ?: return
        if (sessionId in deletedSessions) return
        val updatedAt = value.number("updated_at") ?: return
        val current = sessions[sessionId]
        if (current != null && (current.number("updated_at") ?: 0) > updatedAt) return
        val project = value["project"] as? JsonObject ?: return
        sessions[sessionId] = buildJsonObject {
            put("id", sessionId)
            put("title", value.text("title") ?: "Session")
            put("updated_at", updatedAt)
            put("status", value.text("status") ?: "idle")
            if (value["archived"]?.jsonPrimitive?.contentOrNull == "true") put("archived", true)
            put("project_id", project.text("id") ?: return)
            put("project_name", project.text("name") ?: return)
            put("cwd", project.text("cwd") ?: return)
            put("provider", value.text("provider") ?: return)
            value.text("model")?.let { put("model", it) }
            value.text("reasoning_effort")?.let { put("reasoning_effort", it) }
            put("extensions", value["extensions"] as? JsonArray ?: JsonArray(emptyList()))
        }
        drainPendingDeltas(sessionId)
    }

    private fun applyUpdate(value: JsonObject, queueIfMissing: Boolean) {
        val sessionId = value.text("session_id") ?: return
        val current = sessions[sessionId] ?: run {
            if (queueIfMissing) queuePendingDelta(sessionId, value)
            return
        }
        val updatedAt = value.number("updated_at") ?: return
        if ((current.number("updated_at") ?: 0) > updatedAt) return
        val next = current.toMutableMap()
        next["updated_at"] = JsonPrimitive(updatedAt)
        value.text("title")?.let { next["title"] = JsonPrimitive(it) }
        value.text("provider")?.let { next["provider"] = JsonPrimitive(it) }
        applyNullableText(value, next, "model")
        applyNullableText(value, next, "reasoning_effort")
        (value["extensions"] as? JsonArray)?.let { next["extensions"] = it }
        (value["project"] as? JsonObject)?.let { project ->
            project.text("id")?.let { next["project_id"] = JsonPrimitive(it) }
            project.text("name")?.let { next["project_name"] = JsonPrimitive(it) }
            project.text("cwd")?.let { next["cwd"] = JsonPrimitive(it) }
        }
        sessions[sessionId] = JsonObject(next)
    }

    private fun applyLifecycle(value: JsonObject, queueIfMissing: Boolean) {
        val sessionId = value.text("session_id") ?: return
        val state = value.text("state") ?: return
        val current = sessions[sessionId]
        if (current == null && queueIfMissing) {
            queuePendingDelta(sessionId, value)
            return
        }
        if (state == "deleted") {
            deletedSessions[sessionId] = nativeRevision(value)
            sessions.remove(sessionId)
            return
        }
        current ?: return
        val updatedAt = value.number("updated_at") ?: return
        if ((current.number("updated_at") ?: 0) > updatedAt) return
        sessions[sessionId] = JsonObject(current + buildMap {
            put("updated_at", JsonPrimitive(updatedAt))
            put("status", JsonPrimitive(if (state == "archived") "idle" else state))
            if (state == "archived") put("archived", JsonPrimitive(true))
            else put("archived", JsonPrimitive(false))
        })
    }

    private fun queuePendingDelta(sessionId: String, value: JsonObject) {
        val pending = pendingDeltas.getOrPut(sessionId) { mutableListOf() }
        require(pending.size < MAX_PENDING_DELTAS_PER_SESSION) {
            "Too many Matrix session deltas before root $sessionId"
        }
        pending += value
    }

    private fun drainPendingDeltas(sessionId: String) {
        pendingDeltas.remove(sessionId)
            ?.sortedBy { it.number("updated_at") ?: 0 }
            ?.forEach { value ->
                when (value.text("kind")) {
                    "session_update" -> applyUpdate(value, queueIfMissing = false)
                    "session_lifecycle" -> applyLifecycle(value, queueIfMissing = false)
                }
            }
    }

    private fun applyNullableText(
        source: JsonObject,
        target: MutableMap<String, JsonElement>,
        key: String,
    ) {
        if (key !in source) return
        val text = source.text(key)
        if (text == null) target.remove(key) else target[key] = JsonPrimitive(text)
    }

    @Synchronized
    fun snapshot(): JsonObject? {
        val value = checkpoint ?: return null
        val revision = latestRevision ?: return null
        val workspace = value["workspace"] as? JsonObject ?: return null
        val project = workspace["project"] as? JsonObject ?: return null
        return buildJsonObject {
            put("version", 1)
            put("kind", "gateway_state")
            put("revision", revision.revision)
            put("revision_epoch", revision.epoch)
            put("revision_epoch_generation", revision.generation)
            put("state_version", value.number("state_version") ?: return null)
            put("active_device_count", value.number("active_device_count") ?: return null)
            put("current_session_id", kotlinx.serialization.json.JsonNull)
            put("sessions", JsonArray(sessions.values.sortedByDescending {
                it.number("updated_at") ?: 0
            }))
            put("workspace", buildJsonObject {
                put("project_id", project.text("id") ?: return null)
                put("project_name", project.text("name") ?: return null)
                put("cwd", project.text("cwd") ?: return null)
                put("provider", workspace.text("provider") ?: return null)
                workspace.text("model")?.let { put("model", it) }
                workspace.text("reasoning_effort")?.let { put("reasoning_effort", it) }
                put("permission_mode", workspace.text("permission_mode") ?: "default")
            })
            put("capabilities", value["capabilities"] ?: return null)
        }
    }

    private fun observeRevision(value: JsonObject) {
        val next = nativeRevision(value)
        val current = latestRevision
        if (current == null || next.generation > current.generation) {
            latestRevision = next
            return
        }
        if (next.generation < current.generation) return
        require(next.epoch == current.epoch) {
            "Matrix native events disagree on the revision epoch."
        }
        if (next.revision >= current.revision) latestRevision = next
    }

    private fun nativeRevision(value: JsonObject) = NativeRevision(
            revision = value.number("revision") ?: error("Matrix native revision is missing."),
            epoch = value.text("revision_epoch")
                ?: error("Matrix native revision epoch is missing."),
            generation = value.number("revision_epoch_generation")
                ?: error("Matrix native revision generation is missing."),
        ).also {
            require(it.revision >= 0 && it.generation > 0)
        }

    private fun isOlderThanCheckpoint(value: JsonObject): Boolean {
        val floor = checkpoint?.let(::nativeRevision) ?: return false
        return compareRevisions(nativeRevision(value), floor) < 0
    }

    private fun compareRevisions(left: NativeRevision, right: NativeRevision): Int {
        if (left.generation != right.generation) {
            return left.generation.compareTo(right.generation)
        }
        require(left.epoch == right.epoch) {
            "Matrix native events disagree on the revision epoch."
        }
        return left.revision.compareTo(right.revision)
    }

    private companion object {
        const val MAX_PENDING_DELTAS_PER_SESSION = 256
        val NATIVE_KINDS = setOf(
            "gateway_checkpoint",
            "gateway_revision",
            "session_root",
            "session_update",
            "session_lifecycle",
        )
    }
}

private fun JsonObject.text(key: String): String? =
    this[key]?.jsonPrimitive?.contentOrNull

private fun JsonObject.number(key: String): Long? =
    this[key]?.jsonPrimitive?.longOrNull

File: apps/pwa/app/matrixNativeProjection.ts
import {
  matrixNativeContentSchema,
  type MatrixNativeContent,
} from "@codever/protocol";
import {
  parseGatewayStateExtension,
  type GatewaySessionSummary,
  type GatewayStateSnapshot,
} from "./gatewayState";

type SessionDelta = Extract<
  MatrixNativeContent,
  { kind: "session_update" | "session_lifecycle" }
>;
type NativeRevision = Pick<
  MatrixNativeContent,
  "revision" | "revision_epoch" | "revision_epoch_generation"
>;

const MAX_PENDING_DELTAS_PER_SESSION = 256;

/** Deterministically rebuilds the existing UI model from Matrix-native events. */
export class MatrixNativeProjection {
  private readonly sessions = new Map<string, GatewaySessionSummary>();
  private readonly pendingDeltas = new Map<string, SessionDelta[]>();
  private readonly deletedAt = new Map<string, number>();
  private checkpoint: Extract<MatrixNativeContent, { kind: "gateway_checkpoint" }> | null = null;
  private latestRevision: NativeRevision | null = null;

  apply(input: unknown): GatewayStateSnapshot | null {
    const event = matrixNativeContentSchema.parse(input);
    this.observeRevision(event);
    if (event.kind === "gateway_checkpoint") {
      if (!this.checkpoint || event.state_version >= this.checkpoint.state_version) {
        this.checkpoint = event;
      }
      return this.snapshot();
    }
    if (event.kind === "gateway_revision") return this.snapshot();
    if (event.kind === "session_root") {
      if (this.deletedAt.has(event.session_id)) return this.snapshot();
      const current = this.sessions.get(event.session_id);
      if (!current || event.updated_at >= current.updatedAt) {
        this.sessions.set(event.session_id, {
          id: event.session_id,
          title: event.title,
          updatedAt: event.updated_at,
          status: event.archived ? "archived" : event.status,
          projectId: event.project.id,
          projectName: event.project.name,
          cwd: event.project.cwd,
          provider: event.provider,
          ...(event.model ? { model: event.model } : {}),
          ...(event.reasoning_effort
            ? { reasoningEffort: event.reasoning_effort }
            : {}),
          extensions: event.extensions,
        });
      }
      this.drainPendingDeltas(event.session_id);
      return this.snapshot();
    }
    if (!this.sessions.has(event.session_id)) {
      this.queuePendingDelta(event);
      return this.snapshot();
    }
    this.applyDelta(event);
    return this.snapshot();
  }

  private applyDelta(event: SessionDelta): void {
    if (event.kind === "session_update") {
      const current = this.sessions.get(event.session_id);
      if (!current || event.updated_at < current.updatedAt) return;
      this.sessions.set(event.session_id, {
        ...current,
        updatedAt: event.updated_at,
        ...(event.title !== undefined ? { title: event.title } : {}),
        ...(event.project
          ? {
              projectId: event.project.id,
              projectName: event.project.name,
              cwd: event.project.cwd,
            }
          : {}),
        ...(event.provider !== undefined ? { provider: event.provider } : {}),
        ...(event.extensions !== undefined ? { extensions: event.extensions } : {}),
        ...nullableProperty("model", event.model),
        ...nullableProperty("reasoningEffort", event.reasoning_effort),
      });
      return;
    }
    if (event.state === "deleted") {
      this.deletedAt.set(
        event.session_id,
        Math.max(event.updated_at, this.deletedAt.get(event.session_id) ?? 0),
      );
      this.sessions.delete(event.session_id);
      return;
    }
    const current = this.sessions.get(event.session_id);
    if (!current || event.updated_at < current.updatedAt) return;
    this.sessions.set(event.session_id, {
      ...current,
      updatedAt: event.updated_at,
      status: event.state === "archived" ? "archived" : event.state,
    });
  }

  private queuePendingDelta(event: SessionDelta): void {
    const pending = this.pendingDeltas.get(event.session_id) ?? [];
    if (pending.length >= MAX_PENDING_DELTAS_PER_SESSION) {
      throw new Error(`Too many Matrix session deltas before root ${event.session_id}`);
    }
    pending.push(event);
    this.pendingDeltas.set(event.session_id, pending);
  }

  private drainPendingDeltas(sessionId: string): void {
    const pending = this.pendingDeltas.get(sessionId);
    if (!pending) return;
    this.pendingDeltas.delete(sessionId);
    pending
      .sort((left, right) => left.updated_at - right.updated_at)
      .forEach((event) => this.applyDelta(event));
  }

  applySessionStatus(input: Record<string, unknown>): GatewayStateSnapshot | null {
    if (
      input.kind !== "status" ||
      typeof input.session_id !== "string" ||
      !input.session_id
    ) return this.snapshot();
    const current = this.sessions.get(input.session_id);
    if (!current) return this.snapshot();
    const status =
      input.state === "running" || input.state === "stopping" || input.state === "failed"
        ? input.state
        : "idle";
    const activityPhase =
      input.activity_phase === "starting" ||
      input.activity_phase === "working" ||
      input.activity_phase === "stopping" ||
      input.activity_phase === "idle" ||
      input.activity_phase === "failed"
        ? input.activity_phase
        : undefined;
    this.sessions.set(input.session_id, {
      ...current,
      status,
      ...(activityPhase ? { activityPhase } : {}),
    });
    return this.snapshot();
  }

  snapshot(): GatewayStateSnapshot | null {
    const checkpoint = this.checkpoint;
    if (!checkpoint) return null;
    const revision = this.latestRevision ?? checkpoint;
    return parseGatewayStateExtension({
      version: 1,
      kind: "gateway_state",
      state_version: checkpoint.state_version,
      revision: revision.revision,
      revision_epoch: revision.revision_epoch,
      revision_epoch_generation: revision.revision_epoch_generation,
      active_device_count: checkpoint.active_device_count,
      current_session_id: null,
      sessions: [...this.sessions.values()]
        .sort((left, right) => right.updatedAt - left.updatedAt)
        .map((session) => ({
          id: session.id,
          title: session.title,
          updated_at: session.updatedAt,
          status: session.status === "archived" ? "idle" : session.status,
          ...(session.activityPhase ? { activity_phase: session.activityPhase } : {}),
          ...(session.status === "archived" ? { archived: true } : {}),
          project_id: session.projectId,
          project_name: session.projectName,
          cwd: session.cwd,
          provider: session.provider,
          ...(session.model ? { model: session.model } : {}),
          ...(session.reasoningEffort
            ? { reasoning_effort: session.reasoningEffort }
            : {}),
          extensions: session.extensions,
        })),
      workspace: {
        project_id: checkpoint.workspace.project.id,
        project_name: checkpoint.workspace.project.name,
        cwd: checkpoint.workspace.project.cwd,
        provider: checkpoint.workspace.provider,
        ...(checkpoint.workspace.model ? { model: checkpoint.workspace.model } : {}),
        ...(checkpoint.workspace.reasoning_effort
          ? { reasoning_effort: checkpoint.workspace.reasoning_effort }
          : {}),
        permission_mode: checkpoint.workspace.permission_mode,
      },
      capabilities: checkpoint.capabilities,
    });
  }

  private observeRevision(event: NativeRevision): void {
    const current = this.latestRevision;
    if (!current || event.revision_epoch_generation > current.revision_epoch_generation) {
      this.latestRevision = revisionOf(event);
      return;
    }
    if (event.revision_epoch_generation < current.revision_epoch_generation) return;
    if (event.revision_epoch !== current.revision_epoch) {
      throw new Error("Matrix native events disagree on the revision epoch.");
    }
    if (event.revision >= current.revision) this.latestRevision = revisionOf(event);
  }
}

function revisionOf(value: NativeRevision): NativeRevision {
  return {
    revision: value.revision,
    revision_epoch: value.revision_epoch,
    revision_epoch_generation: value.revision_epoch_generation,
  };
}

function nullableProperty(
  key: "model" | "reasoningEffort",
  value: string | null | undefined,
): Partial<Pick<GatewaySessionSummary, "model" | "reasoningEffort">> {
  if (value === undefined) return {};
  return value === null ? { [key]: undefined } : { [key]: value };
}
